### ISSUE TITLE
[Possible Bug] PhpRedis keeps crashing PHP when a key expires

### ISSUE THREAD

**Issue by ProjectCleverWeb** [phpredis/phpredis#1432]
### Expected behaviour
The millisecond a key expires `get()` should return `FALSE`.

### Actual behaviour
The millisecond a key expires and `get()` is called, PHP suddenly exits without showing/logging an error (or running shutdown behavior)

### I'm seeing this behaviour on
- OS: Amazon Linux (fork of RHEL/Fedora)
- Redis: 4.0.10 via AWS ElastiCache (Clustered, 1 Shard/2 Nodes)
- PHP: 7.0.31 (Running as PHP-FPM)
- phpredis: 3.1.1-1.1.amzn1

### Steps to reproduce, backtrace or example script
Attempt to `get()` a key several times at about the time it is set to expire. I know that it is specifically `get()` that's failing, because I had PHP write to a text file before and after each interaction with Redis, including connecting to Redis.

In my particular case, the expire is set to 10 seconds and if a request happens exactly 10 seconds later, PHP-FPM exits without showing an error (or running shutdown behavior) and resets the connection to Nginx, which nginx then responds with a 502 Bad Gateway error. The API that I am integrating Redis into handles about 5-10 requests per second during peak hours, so this would become a common error.

The real version of the below code is a class called by an endpoint, but for simplicity's sake I have removed code that isn't related to PhpRedis:

```php
function cache_store(string $key, $value, int $lifespan = 10) {
	// [not shown] checking if we successfully connected to Redis, if both $key
	// and $lifespan are valid, and if we can serialize $value
	
	// $connection is an instance of RedisCluster
	global $connection;
	
	return $connection->setEx($key, $lifespan, serialize($value));
}

function cache_exists(string $key) {
	// [not shown] checking if we successfully connected to Redis and if $key is
	// valid
	
	// $connection is an instance of RedisCluster
	global $connection;
	
	return $connection->exists($key);
}

function cache_read(string $key) {
	// Checks if connected & valid key & exists
	if (!cache_exists($key)) {
		return NULL;
	}
	
	// $connection is an instance of RedisCluster
	global $connection;
	
	$value = $connection->get($key);
	if (!$value) {
		return NULL;
	}
	return unserialize($value);
}

try {
	$connection = new \RedisCluster(NULL, array('redis.aws.example.com'),1,1);
} catch (\Exception $e) {
	echo $e->getMessage();
	exit;
}
$cache_key = 'core:time';
$time      = cache_read($cache_key);
if (is_null($time)) {
	$time = time();
	cache_store($cache_key, $time, 10);
}

var_dump($time);
```

### I've checked
- [ X ] There is no similar issue from other users
- [ **not sure** ] Issue isn't fixed in `develop` branch

**Comment by ProjectCleverWeb** [comment]
After further research, this is still happening but it's probably not related to expired keys. My first few tests just happen to occur at about the time of expire. I was able to recreate this issue where it happened only 2 seconds after a key with a 10 second expire was created.

**Comment by michael-grunder** [comment]
This script doesn't cause any issues for me.  I also modified it to run in a tight loop and can't get Phpredis to crash either.

I'm compiling php 7.0.1 in order to test it against the same version of PHP but I'm guessing it's not going to break.

Are you able to replicate this via a cli script?  I think the biggest hurdle here is going to be replicating the bug.

**Comment by ProjectCleverWeb** [comment]
I haven't tried via CLI yet, but I will soon.

It is very intermittent, it is possible that it's related to the performance of the redis cluster, or possibly even something to do with my specific deployment?

However, this has happened when the cluster had 0 load. In fact, this isn't deployed to production yet so I am the only user actually connecting to Redis.

I will keep testing today and see if I can narrow down a specific cause, or at least a consistent way to recreate the issue. I will also try resizing and redeploying the cluster to see if the issue goes away then.

**Comment by yatsukhnenko** [comment]
@ProjectCleverWeb please make your tests with the latest stable release of phpredis

**Comment by ProjectCleverWeb** [comment]
I actually literally just got the package conflicts resolved and tested 4.1.1

The issues **seems** to be fixed! 😄 I will close this issue in the next hour or 2 if I am unable to replicate the issue.

Thanks for all your help @michael-grunder and @yatsukhnenko

**Comment by michael-grunder** [comment]
That's interesting about php-fpm.  Honestly it's not really my area of expertise but in theory it really shouldn't freeze anything.  

You'll need to cycle the process but I don't know why (or even how) pecl would cause php-fpm to freeze.
